### PR TITLE
Fix #9385: Add a workflow to check if artifacts have changed

### LINKED ISSUE #9385: [CT-3555] Add detect artifact changes workflow
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

We should detect whether a PR makes changes to an existing artifact schema, and fail CI if this is the case. https://github.com/marketplace/actions/prevent-file-change might do the trick!

Or, if we don't want to use an (overly-complex) public action, could do something manually: 
* check which files changed, and fail if any of those are modified
* allow new files in dbt/artifacts

### Acceptance criteria

* CI fails if any file is modified in dbt/artifacts
* failure message provides details 
* CI failure should either: not block merge, or is made to pass by setting a label (e.g. artifact_minor_upgrade)

Future work could be only failing the check if _breaking_ changes are introduced. Until then, we can manually inspect the changes and merge on red if the change is deemed safe. 

Additionally, tests in dbt/artifacts should provide signal as to whether a breaking change is made to an existing artifact because existing upgrade/downgrade logic would fail. 

### Impact to Other Teams

N/A - CI work in dbt-core

### PR DESCRIPTION
**Pull request by aranke**
resolves #9385 

### Problem

It isn't always clear when a file in the `core/dbt/artifacts` directory has been changed.

### Solution

This PR adds a CI check that:
1. Fails if any file in `core/dbt/artifacts` has been changed
2. Can be overridden by adding the `artifact_minor_upgrade` tag to the PR.

This solution currently uses `git diff` and doesn't rely on any external actions.

Run history: 
1. Example failure: https://github.com/dbt-labs/dbt-core/actions/runs/7866653246/job/21461212098?pr=9553
2. Example success (with label): https://github.com/dbt-labs/dbt-core/actions/runs/7866663319/job/21461233620?pr=9553
3. All runs: https://github.com/dbt-labs/dbt-core/actions/workflows/check-artifact-changes.yml

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

File: .github/workflows/check-artifact-changes.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            echo "$CHANGED_FILES"
+          else
+            echo "CHANGED=false" >> $GITHUB_ENV
+          fi
+
+      - name: Fail CI if artifacts have changed
+        if: env.CHANGED == 'true' && !contains(github.event.pull_request.labels.*.name, 'artifact_minor_upgrade')
+        run: |
+          echo "CI failure: Artifact changes checked in core/dbt/artifacts directory."
+          echo "To bypass this check, add the 'artifact_minor_upgrade' label to the PR."
```
Comment (MichelleArk) [inline .github/workflows/check-artifact-changes.yml:36]: Could you update this message to include a link to the [breaking changes](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/artifacts/README.md#breaking-changes) section of the dbt/artifacts/README.md? 

Perhaps something like: "To bypass this check, confirm that the change is not breaking (link) and add the 'artifact_minor_upgrade' label to the PR."

File: .github/workflows/check-artifact-changes.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      - name: Checkout code
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Check for changes in core/dbt/artifacts
+        id: check_changes
+        run: |
+          # Check for changes in dbt/artifacts
+          CHANGED_FILES=$(git diff --name-only origin/${GITHUB_BASE_REF} origin/${GITHUB_HEAD_REF} | grep 'core/dbt/artifacts/')
```
Comment (MichelleArk) [inline .github/workflows/check-artifact-changes.yml:23]: Could we exclude matching `core/dbt/artifacts/README.md` as part of this search?

**Review by MichelleArk — APPROVED** [review]
couple small comments re: linking to and ignoring changes to core/dbt/artifacts/README.md, but otherwise looks great! thank you!

File: .github/workflows/check-artifact-changes.yml (hunk 2 above)
Comment (aranke) [reply .github/workflows/check-artifact-changes.yml:23]: We're ignoring all `*.md` files, so we should be fine:

https://github.com/dbt-labs/dbt-core/pull/9553/files#diff-a7bcbf79c3b395a73ebf1356d7c6a129c82f2f01d385cfbf82cfc4af21f5b3c0R6

**Pull request merged**